### PR TITLE
GAP-2060 - Enhanced block

### PR DESCRIPTION
**Pull request by john-tco**
## Description

Ticket # GAP-2060

- adds a endpoint to validate the admin session
- updates jwtService.verifyToken call to validate roles
- adds an endpoint to get the user data attached to the jwt (super-admin fe uses this to check who they are blocking / deleting)
- Stops super admins from blocking themselves

Related prs:

- https://github.com/cabinetoffice/gap-find-admin-backend/pull/29
- https://github.com/cabinetoffice/gap-find-apply-web/pull/68 
- https://eu-west-2.console.aws.amazon.com/codesuite/codecommit/repositories/GAP-infra/pull-requests/96/details?region=eu-west-2

Todos:
- add more unit tests

## Type of change

Please check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [x] Unit Test

- [ ] Integration Test (if applicable)

- [ ] End to End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

# Checklist:

- [ ] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [x] I have performed a self-review of my code.
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation where applicable.
- [ ] I have ran cypress tests and they all pass locally.

### CODE REVIEW

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 1]
```diff
@@ -166,4 +176,18 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
         thirdPartyAuthToken.setMaxAge(0);
         response.addCookie(thirdPartyAuthToken);
     }
+
+    public void validateRoles(List<Role> dbRoles, String roles) {
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:180]: Same here for the naming of dbRoles and roles. 

It should be clearer that one is the user associated roles stored in the db and the other is roles associated with the payload

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 2]
```diff
@@ -166,4 +176,18 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
...
+        final List<String> rolesList = removeSquareBracketsAndTrim(Arrays.asList(roles.split(",")));
+        final List<String> dbRolesList = dbRoles.stream().map(role -> roleMapper.roleToRoleDto(role).getName()).collect(Collectors.toList());
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:182]: I think you can replace :

`.collect(Collectors.toList());
`
with

`.toList();`

Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Nice, yeah that works.

File: src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java [hunk 3]
```diff
@@ -147,4 +160,22 @@ public JwtPayload decodeTheTokenPayloadInAReadableFormat(DecodedJWT jwt) {
...
+        if (customJWTCookie == null || customJWTCookie.getValue() == null) {
+            throw new UnauthorizedException("No JWT token provided");
+        }
+        final DecodedJWT decodedJwt = decodedJwt(customJWTCookie.getValue());
+        final JwtPayload payload = decodeTheTokenPayloadInAReadableFormat(decodedJwt);
+        return  userRepository.findBySub(payload.getSub()).get();
+    }
+
+    public JwtPayload validateRolesInThePayload(JwtPayload payload) throws UnauthorizedException {
+        final List<Role> dbRoles = oneLoginUserService.getUserBySub(payload.getSub()).getRoles();
```
[the remark this answers is not in the document]
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java]: Renamed that. I actually had `userRoles` in an earlier commit lol

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 4]
```diff
@@ -166,4 +176,18 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
...
+
+        if(!new HashSet<>(dbRolesList).containsAll(rolesList)){
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Can you extract the hashset instantiation out for readability?

HashSet<String> userStoredRoles = new HashSet<>(dbRolesList);

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 5]
```diff
@@ -166,4 +175,19 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
         thirdPartyAuthToken.setMaxAge(0);
         response.addCookie(thirdPartyAuthToken);
     }
+
+    public void validateRoles(List<Role> userRoles, String payloadRoles) {
+        final List<String> formattedPayloadRoles = removeSquareBracketsAndTrim(Arrays.asList(payloadRoles.split(",")));
+        final List<String> formattedUserRoles = userRoles.stream().map(role -> roleMapper.roleToRoleDto(role).getName()).toList();
+        HashSet userStoredRoles = new HashSet<>(formattedUserRoles);
```
Comment (ryan-tco) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:182]: Why not create formattedUserRoles as a set with collectors.toSet to avoid the need to instantiate a new set

File: src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java [hunk 6]
```diff
@@ -87,6 +97,9 @@ public boolean isTokenValid(final String customJwt) {
         } catch (JOSEException exception) {
             log.error("isTokenValid failed", exception);
             throw new RuntimeException(exception);
+        } catch (UnauthorizedException exception) {
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java:100]: There is no unit test coverage for this

File: src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java [hunk 7]
```diff
@@ -147,4 +160,22 @@ public JwtPayload decodeTheTokenPayloadInAReadableFormat(DecodedJWT jwt) {
...
     }
+
+    public User getUserFromJwt(final HttpServletRequest request) {
+        final Cookie customJWTCookie = WebUtils.getCookie(request, userServiceCookieName);
+        if (customJWTCookie == null || customJWTCookie.getValue() == null) {
+            throw new UnauthorizedException("No JWT token provided");
+        }
+        final DecodedJWT decodedJwt = decodedJwt(customJWTCookie.getValue());
+        final JwtPayload payload = decodeTheTokenPayloadInAReadableFormat(decodedJwt);
+        return  userRepository.findBySub(payload.getSub()).get();
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java:171]: userRepository.findBySub returns an optional so should check if it's present before getting it

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 8]
```diff
@@ -184,4 +193,19 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
         thirdPartyAuthToken.setMaxAge(0);
         response.addCookie(thirdPartyAuthToken);
     }
+
+    public void validateRoles(List<Role> userRoles, String payloadRoles) throws UnauthorizedException {
+        final List<String> formattedPayloadRoles = removeSquareBracketsAndTrim(Arrays.asList(payloadRoles.split(",")));
+        final List<String> formattedUserRoles = userRoles.stream().map(role -> roleMapper.roleToRoleDto(role).getName()).toList();
+        HashSet userStoredRoles = new HashSet<>(formattedUserRoles);
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:200]: HashSet<String> *

File: src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java [hunk 9]
```diff
@@ -75,6 +81,12 @@ public class LoginControllerV2 {
     @Value("${feature.onelogin.migration.enabled}")
     public String migrationEnabled;
 
+    @GetMapping("/validateAdminSession")
+    public ResponseEntity<Boolean> validateAdminRoles(@RequestHeader("emailaddress") String emailAddress, @RequestHeader("roles") String roles){
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java:85]: Any reason these are requestHeaders and not request parameters?

File: src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java [hunk 10]
```diff
@@ -29,8 +30,18 @@ public class UserController {
     private final OneLoginUserService oneLoginUserService;
     private final DepartmentService departmentService;
     private final RoleService roleService;
+    private final CustomJwtServiceImpl jwtService;
     private final SecretAuthService secretAuthService;
 
+    @GetMapping("/userFromJwt")
+    public ResponseEntity<UserDto> getUserFromJwt(HttpServletRequest httpRequest) {
+        if (!roleService.isSuperAdmin(httpRequest)) {
+            throw new ForbiddenException();
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java:39]: Can write unit test to cover this scenario

File: src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java [hunk 11]
```diff
@@ -94,6 +105,12 @@ public ResponseEntity<String> updateRoles(HttpServletRequest httpRequest, @Reque
             throw new ForbiddenException();
         }
 
+        boolean requestToBlockUser = roleIds.size() == 0;
+
+        if (requestToBlockUser && id == jwtService.getUserFromJwt(httpRequest).getGapUserId()){
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java:110]: you should use .equals() instead of ==

File: src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java [hunk 12]
```diff
@@ -147,4 +160,21 @@ public JwtPayload decodeTheTokenPayloadInAReadableFormat(DecodedJWT jwt) {
...
     }
+
+    public Optional<User> getUserFromJwt(final HttpServletRequest request) {
+        final Cookie customJWTCookie = WebUtils.getCookie(request, userServiceCookieName);
+        if (customJWTCookie == null || customJWTCookie.getValue() == null) {
+            throw new UnauthorizedException("No JWT token provided");
+        }
+        final DecodedJWT decodedJwt = decodedJwt(customJWTCookie.getValue());
+        final JwtPayload payload = decodeTheTokenPayloadInAReadableFormat(decodedJwt);
+        return  userRepository.findBySub(payload.getSub());
```
Comment (jgunnCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java:171]:
```suggestion
        return userRepository.findBySub(payload.getSub());
```

File: src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java (hunk 9 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/web/LoginControllerV2.java]: It came through as `emailaddress` when it was passed via headers in camelcase, have switched to query params & camel case does suit now

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 4 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Using a Set<String> for this now

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 5 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Good shout :)

File: src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java (hunk 6 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java:100]: Added that

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 8 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Have switched to `Set<String>`

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 13]
```diff
@@ -184,4 +193,19 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
...
+
+        if(!userStoredRoles.containsAll(formattedPayloadRoles)){
+            throw new UnauthorizedException("Roles in payload do not match roles in database");
+        }
+    }
+
+    public void validateAdminSession(String emailAddress, String roles) throws UnauthorizedException {
+        final List<Role> dbRoles = userRepository.findByEmailAddress(emailAddress).get().getRoles();
```
[the remark this answers is not in the document]
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:210]: Now throwing an error when we can't get the optional user

File: src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java (hunk 10 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java:40]: Added a few to cover this controller

File: src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java (hunk 11 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java]: Done

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 14]
```diff
@@ -184,4 +194,20 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
...
         response.addCookie(thirdPartyAuthToken);
     }
+
+    public void validateRoles(List<Role> userRoles, String payloadRoles) throws UnauthorizedException {
+        final List<String> formattedPayloadRoles = removeSquareBracketsAndTrim(Arrays.asList(payloadRoles.split(",")));
+        final Set<String> formattedUserRoles = userRoles.stream()
+                .map(role -> roleMapper.roleToRoleDto(role).getName())
+                .collect(Collectors.toSet());
+
+        if(!formattedUserRoles.containsAll(formattedPayloadRoles)){
```
Comment (westdom) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:204]: This check doesn't cover all scenarios. Just ensures formattedPayloadRoles is a subset of formattedUserRoles. What about the inverse scenario?

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 15]
```diff
@@ -184,4 +194,20 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
...
+            throw new UnauthorizedException("Roles in payload do not match roles in database");
+        }
+    }
+
+    public void validateAdminSession(String emailAddress, String roles) throws UnauthorizedException {
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:209]: Any reason you have throws UnauthorizedException in the method signature? i think you can remove this from here and every method that calls it

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:210]: Screenshot 2023-09-12 at 09 51 06

Doesn't look like this has been pushed

File: src/main/java/gov/cabinetofice/gapuserservice/util/HelperUtils.java [hunk 16]
```diff
@@ -57,6 +59,15 @@ public static String asJsonString(final Object obj) {
         }
     }
 
+    public static List<String> removeSquareBracketsAndTrim(List<String> inputList) {
+        List<String> cleanedList = new ArrayList<>();
+        for (String input : inputList) {
+            String cleanedString = input.replace("[", "").replace("]", "");
+            cleanedList.add(cleanedString.trim());
+        }
+        return cleanedList;
```
Comment (westdom) [inline src/main/java/gov/cabinetofice/gapuserservice/util/HelperUtils.java:68]: I'm a sucker for streams
```suggestion
        return inputList.stream()
                .map(input -> input.replace("[", "").replace("]", "").trim())
                .collect(Collectors.toList());
```

File: src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java (hunk 10 above)
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java:40]: Screenshot 2023-09-12 at 09 52 41

No test to cover what happens when user is empty?

File: src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java (hunk 7 above)
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java]: Screenshot 2023-09-12 at 09 56 03

This method still has no coverage?

File: src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java [hunk 17]
```diff
@@ -94,6 +110,16 @@ public ResponseEntity<String> updateRoles(HttpServletRequest httpRequest, @Reque
             throw new ForbiddenException();
         }
 
+        boolean requestToBlockUser = roleIds.size() == 0;
```
Comment (westdom) [inline src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java:113]: IsARequestToBlockUser

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 14 above)
Comment (westdom) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:204]: Ideally add unit tests that would capture this scenario too. Along with any other edge cases that arent already tested

File: src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java (hunk 7 above)
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java]: Screenshot 2023-09-12 at 11 39 29

Still no coverage for this method?

**Review by dylanwrightCO — COMMENTED** [review]
Just some general comments:

line length - Ideally we don't go over the white line on the right seen on IntelliJ.  Keeps the code cleaner and easier to review

code coverage - You should run all test with coverage on IntelliJ and it will highlight the lines covered in green and those not in red.  It's important that we aim for 100% coverage when we can.

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 15 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Leftover from when `validateRoles` above wasn't extracted. Have removed

File: src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java (hunk 10 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/web/UserController.java:40]: added that

File: src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java (hunk 7 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/jwt/impl/CustomJwtServiceImpl.java]: pushed that now

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 14 above)
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:202]: If both collections are a set you can check if they are equal to each other instead of containsAll.  As Dom said we need unit tests to cover all scenarios

File: src/test/java/gov/cabinetofice/gapuserservice/service/jwt/CustomJwtServiceImplTest.java [hunk 18]
```diff
@@ -251,4 +300,59 @@ void ReturnsACustomJwt() {
...
+        @Test
+        void testGetUserFromJwtThrowsUnauthorizedWithNullJwt() {
+            HttpServletRequest mockRequest = mock(HttpServletRequest.class);
+            assertThrows(UnauthorizedException.class, () -> serviceUnderTest.getUserFromJwt(mockRequest));
+        }
+
+        @Test
+        void testGetUserFromJwtWithNullValueInJwt() {
+            HttpServletRequest mockRequest = mock(HttpServletRequest.class);
+            mockedWebUtils.when((MockedStatic.Verification) WebUtils.getCookie(mockRequest, null)).thenReturn(null);
```
[the remark this answers is not in the document]
Comment (john-tco) [reply src/test/java/gov/cabinetofice/gapuserservice/service/jwt/CustomJwtServiceImplTest.java]: It was null because I wasn't able to mock the env var we use for the name of the cookie. Have amended that now

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Have added that

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Sure, using that now

File: src/main/java/gov/cabinetofice/gapuserservice/util/HelperUtils.java (hunk 16 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/util/HelperUtils.java]: sure, happy with that too

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 14 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: added 4 tests for this which should cover everything

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Could be simplified to this using orElseThrows():

`final List<Role> userRoles = userRepository.findByEmailAddress(emailAddress)
                .orElseThrow(() -> new InvalidRequestException("Could not get user from emailAddress")).getRoles();`

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 15 above)
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Doesn't look like it's been removed?

The method above also doesn't need to have it in the method signature

Screenshot 2023-09-13 at 14 48 40

File: src/test/java/gov/cabinetofice/gapuserservice/service/jwt/CustomJwtServiceImplTest.java [hunk 19]
```diff
@@ -118,6 +141,30 @@ void ReturnsFalse_IfInvalid() {
...
+            final Verification spiedVerification = spy(verification);
+            ReflectionTestUtils.setField(serviceUnderTest, "oneLoginEnabled", true);
+
+            try (MockedStatic<Algorithm> staticAlgorithm = Mockito.mockStatic(Algorithm.class)) {
+                staticAlgorithm.when(() -> RSA256(any(), any())).thenReturn(mockAlgorithm);
+                try (MockedStatic<JWT> staticJwt = Mockito.mockStatic(JWT.class)) {
+                    staticJwt.when(() -> require(any())).thenReturn(spiedVerification);
+                    staticJwt.when(() -> decode(any())).thenCallRealMethod();
+                    when(spiedVerification.build()).thenReturn(mockedJwtVerifier);
+                    User testUser = User.builder().roles(List.of(Role.builder().name(RoleEnum.FIND).id(1).build(), Role.builder().name(RoleEnum.SUPER_ADMIN).id(4).build(), Role.builder().name(RoleEnum.ADMIN).id(3).build(), Role.builder().name(RoleEnum.APPLICANT).id(2).build())).loginJourneyState(LoginJourneyState.USER_READY).build();
```
Comment (dylanwrightCO) [inline src/test/java/gov/cabinetofice/gapuserservice/service/jwt/CustomJwtServiceImplTest.java:156]: line length...

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (dylanwrightCO) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Can you update the test names for this method as they reference the old name

File: src/test/java/gov/cabinetofice/gapuserservice/web/UserControllerTest.java [hunk 20]
```diff
@@ -86,11 +112,58 @@ void shouldReturnChangeDepartmentPageDtoWhenValidIdIsGiven() {
...
+        final HttpServletRequest httpRequest = mock(HttpServletRequest.class);
+        when(roleService.isSuperAdmin(httpRequest)).thenReturn(true);
+        when(customJwtService.getUserFromJwt(httpRequest)).thenReturn(Optional.empty());
+
+        assertThrows(InvalidRequestException.class, () -> controller.deleteUser(httpRequest, 1));
+    }
+
+
+    @Test
+    public void testGetUserFromJwt() throws Exception {
```
[the remark this answers is not in the document]
Comment (john-tco) [reply src/test/java/gov/cabinetofice/gapuserservice/web/UserControllerTest.java]: sure, removed that

File: src/test/java/gov/cabinetofice/gapuserservice/service/jwt/CustomJwtServiceImplTest.java (hunk 19 above)
Comment (john-tco) [reply src/test/java/gov/cabinetofice/gapuserservice/service/jwt/CustomJwtServiceImplTest.java]: fixed

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 15 above)
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Fixed

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Added that

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (john-tco) [reply src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java]: Sure

**Review by dylanwrightCO — DISMISSED** [review]

File: src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java [hunk 21]
```diff
@@ -184,4 +191,22 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
         thirdPartyAuthToken.setMaxAge(0);
         response.addCookie(thirdPartyAuthToken);
     }
+
+    public void validateRoles(List<Role> userRoles, String payloadRoles) {
+        final Set<String> formattedPayloadRoles = removeSquareBracketsAndTrim(Arrays.asList(payloadRoles
```
Comment (dominicwest) [inline src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java:196]: Would rather the logic that splits out the string into an array lives in this helper function, but not overly fussed

**Review by dominicwest — APPROVED** [review]

**Review by dominicwest — DISMISSED** [review]

**Review by dylanwrightCO — APPROVED** [review]
